### PR TITLE
Optimize memory consumption of Gradle dependency tree plugin

### PR DESCRIPTION
**Pull request by oheger-bosch**
This PR addresses #3321 on side of the Gradle plugin used to construct a dependency tree.

The plugin used to create a large number of duplicated Dependency objects. The PR applies a similar optimization than #3502; it caches the Dependency objects and reuses them throughout the dependency graph whenever possible.

Interestingly, this optimization also caused an unexpected speed improvement - probably because the dependency graph now is not traversed again and again for each scope. The time to analyse the [Focus Android](https://github.com/mozilla-mobile/focus-android.git) project on my machine went down from over 30 minutes to less than 4 minutes.

### CODE REVIEW

**Comment by sschuberth** [comment]
> The time to analyse the [Focus Android](https://github.com/mozilla-mobile/focus-android.git) project on my machine went down from over 30 minutes to less than 4 minutes.

That sounds fantastic!

Is my understanding correct that this can be mgered independently of #3502?

**Comment by oheger-bosch** [comment]
> Is my understanding correct that this can be merged independently of #3502?

Yes, this is correct. The both PRs do not depend on each other.

**Review by mnonnenmacher — DISMISSED** [review]

File: analyzer/src/main/resources/scripts/init.gradle [hunk 1]
```diff
@@ -138,22 +138,21 @@ class AbstractDependencyTreePlugin<T> implements Plugin<T> {
...
-
-            def coordinates = displayName.split(':')
-
-            if (coordinates.length == 3) {
-                return new DependencyImpl(groupId: coordinates[0], artifactId: coordinates[1], version: coordinates[2],
-                        dependencies: dependencies, error: error?.toString(), warning: warning?.toString())
-            }
-
-            return new DependencyImpl(groupId: '<unknown>', artifactId: displayName.replace(':', '_'),
+            def coordinates = displayNameToCoordinates(displayName)
```
Comment (sschuberth) [inline analyzer/src/main/resources/scripts/init.gradle:154]: Could we maybe split out changes (like this refactoring) that are not strictly related to the main topic to preceding commits? This make the diff for the important change smaller and thus easier to review.

(This might also be applicable to other functions below, but you know better than me whether that's the case.)

File: analyzer/src/main/resources/scripts/init.gradle [hunk 2]
```diff
@@ -298,7 +333,7 @@ class AbstractDependencyTreePlugin<T> implements Plugin<T> {
             }
         }
 
-        private String collectCauses(Throwable throwable) {
+        private static String collectCauses(Throwable throwable) {
```
Comment (sschuberth) [inline analyzer/src/main/resources/scripts/init.gradle:336]: Nit: Split out to separate commit.

File: analyzer/src/main/resources/scripts/init.gradle [hunk 3]
```diff
@@ -225,14 +224,50 @@ class AbstractDependencyTreePlugin<T> implements Plugin<T> {
                     repositories, errors.unique(), warnings.unique())
         }
 
-        Dependency parseDependency(DependencyResult dependencyResult, Project project,
-                                   Set<ResolvedArtifact> resolvedArtifacts, List<String> parents) {
+        /**
+         * Returns a Dependency for the given DependencyResult. The function checks whether there is already a
+         * Dependency instance in the cache compatible with the result. If this is not the case, parseDependency()
+         * is called to generate a new instance.
+         * @param dependencyResult represents the package to be processed
```
Comment (sschuberth) [inline analyzer/src/main/resources/scripts/init.gradle:231]: Nit: "Blank" comment line (with just `*`) above.

File: analyzer/src/main/resources/scripts/init.gradle [hunk 4]
```diff
@@ -225,14 +224,50 @@ class AbstractDependencyTreePlugin<T> implements Plugin<T> {
...
+
+            dependency = parseDependency(dependencyResult, project, resolvedArtifacts, visited)
+            dependencies[dependencyId].add(dependency)
+            return dependency
+        }
+
+        /**
+         * Creates a new Dependency object to represent the given DependencyResult. The different types of
+         * dependencies are evaluated. The (transitive) dependencies of this dependency are processed recursively.
+         * @param dependencyResult represents the package to be processed
```
Comment (sschuberth) [inline analyzer/src/main/resources/scripts/init.gradle:257]: Nit: "Blank" comment line (with just `*`) above.

File: analyzer/src/main/resources/scripts/init.gradle [hunk 5]
```diff
@@ -307,6 +342,116 @@ class AbstractDependencyTreePlugin<T> implements Plugin<T> {
             }
             return result
         }
-    }
 
+        /**
+         * Checks whether the given dependency and the dependency result spawn the same dependency trees. This is
+         * used to test whether the dependency result can be represented by this dependency.
+         * @param dependency the dependency
```
Comment (sschuberth) [inline analyzer/src/main/resources/scripts/init.gradle:349]: Nit: "Blank" comment line (with just `*`) above.

File: analyzer/src/main/resources/scripts/init.gradle [hunk 6]
```diff
@@ -307,6 +342,116 @@ class AbstractDependencyTreePlugin<T> implements Plugin<T> {
...
+            return dependencyMap.every { entry ->
+                def matchingResult = resultMap[entry.key]
+                matchingResult != null && dependencyTreeEquals(entry.value as Dependency,
+                        matchingResult as DependencyResult, project, nextVisited)
+            }
+        }
+
+        /**
+         * Generates a unique string identifier for a DependencyResult. This is used to de-duplicate dependencies.
+         * @param dependencyResult the dependency result
```
Comment (sschuberth) [inline analyzer/src/main/resources/scripts/init.gradle:381]: Nit: "Blank" comment line (with just `*`) above.

File: analyzer/src/main/resources/scripts/init.gradle [hunk 7]
```diff
@@ -307,6 +342,116 @@ class AbstractDependencyTreePlugin<T> implements Plugin<T> {
...
+            if (dependencyResult instanceof UnresolvedDependencyResult) {
+                return identifierFromDisplayName(dependencyResult.attempted.displayName)
+            }
+
+            return identifierFromDisplayName(dependencyResult.requested.displayName)
+        }
+
+        /**
+         * Generates a unique identifier for the given Dependency.
+         * @param dependency the Dependency
```
Comment (sschuberth) [inline analyzer/src/main/resources/scripts/init.gradle:408]: Nit: "Blank" comment line (with just `*`) above.

File: analyzer/src/main/resources/scripts/init.gradle [hunk 8]
```diff
@@ -307,6 +342,116 @@ class AbstractDependencyTreePlugin<T> implements Plugin<T> {
...
+         * @return the identifier for this Dependency
+         */
+        private static String identifierFor(Dependency dependency) {
+            return toIdentifier(dependency.groupId, dependency.artifactId, dependency.version)
+        }
+
+        /**
+         * Generates a unique string identifier from a display name. The display name is converted to coordinates,
+         * from which the identifier can be constructed.
+         * @param displayName the display name
```
Comment (sschuberth) [inline analyzer/src/main/resources/scripts/init.gradle:418]: Nit: "Blank" comment line (with just `*`) above.

File: analyzer/src/main/resources/scripts/init.gradle [hunk 9]
```diff
@@ -307,6 +342,116 @@ class AbstractDependencyTreePlugin<T> implements Plugin<T> {
...
+         * @return a unique identifier for this display name
+         */
+        private static String identifierFromDisplayName(String displayName) {
+            def coordinates = displayNameToCoordinates(displayName)
+            return toIdentifier(coordinates[0], coordinates[1], coordinates[2])
+        }
+
+        /**
+         * Splits a display name for a component to an array of groupId, artifactId, and version.
+         * @param displayName the name to split
```
Comment (sschuberth) [inline analyzer/src/main/resources/scripts/init.gradle:428]: Nit: "Blank" comment line (with just `*`) above.

File: analyzer/src/main/resources/scripts/init.gradle [hunk 10]
```diff
@@ -307,6 +342,116 @@ class AbstractDependencyTreePlugin<T> implements Plugin<T> {
...
+            if (coordinates.length == 3) {
+                return coordinates
+            }
+
+            return ['<unknown>', displayName.replace(':', '_'), '']
+        }
+
+        /**
+         * Generate a unique identifier for the component with the given coordinates.
+         * @param group the group
```
Comment (sschuberth) [inline analyzer/src/main/resources/scripts/init.gradle:448]: Nit: "Blank" comment line (with just `*`) above.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: analyzer/src/main/resources/scripts/init.gradle (hunk 1 above)
Comment (oheger-bosch) [reply analyzer/src/main/resources/scripts/init.gradle]: Created a separate commit for the refactoring of this function. The other functions belong to the logic to de-duplicate dependencies; so they are left in the main commit.

File: analyzer/src/main/resources/scripts/init.gradle (hunk 2 above)
Comment (oheger-bosch) [reply analyzer/src/main/resources/scripts/init.gradle:331]: Done

**Review by sschuberth — APPROVED** [review]

**Comment by sschuberth** [comment]
Merging despite the bogus failure in *WindowsAnalyzerTest*.
